### PR TITLE
guest_iommu_test: add new case for guest iommu test

### PR DESCRIPTION
**Pull request by ybduan**
ID: 1702195
Signed-off-by: ybduan <yduan@redhat.com>

### CODE REVIEW

**Comment by ybduan** [comment]
(1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.seabios_iommu.q35: PASS (80.63 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.seabios_iommu.q35: PASS (299.43 s)

File: qemu/cfg/seabios_iommu.cfg [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Please add "iommu=pt intel_iommu=on" to kernel line in host
+- seabios_iommu: install setup image_copy unattended_install.cdrom
+    virt_test_type = qemu
+    type = seabios_iommu
+    only q35
+    no WinXP WinVista Win7 Win8 Win8.1 Win2003
+    no Win2008 Win2008..r2 Win2012 Win2012..r2
+    virtio_dev_disable_legacy = on
+    virtio_dev_disable_modern = off
```
Comment (jingzhao84) [inline qemu/cfg/seabios_iommu.cfg]: @ybduan  do we need to add "disable_legacy, disable_modern"?
I think the default virtio driver now is 1.0

File: qemu/cfg/seabios_iommu.cfg [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Please add "iommu=pt intel_iommu=on" to kernel line in host
+- seabios_iommu: install setup image_copy unattended_install.cdrom
+    virt_test_type = qemu
+    type = seabios_iommu
+    only q35
+    no WinXP WinVista Win7 Win8 Win8.1 Win2003
+    no Win2008 Win2008..r2 Win2012 Win2012..r2
```
Comment (jingzhao84) [inline qemu/cfg/seabios_iommu.cfg]: @ybduan  why disable above windows guest? I think we only need to disable WinXP WinVista Win2003

File: qemu/cfg/seabios_iommu.cfg [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    blk_extra_params_stg1 = "serial=TARGET_DISK1"
+    drive_format_stg0 = virtio
+    drive_format_stg1 = scsi-hd
+    virtio_scsi:
+        drive_bus_stg1 += 1
+    nics = "nic0"
+    nic_model_nic0 = virtio
+    device_id_nic0 = nic0
+    Linux:
+        enable_guest_iommu = yes
```
Comment (jingzhao84) [inline qemu/cfg/seabios_iommu.cfg]: @ybduan 
There have two scenarios for iommu in the guest
1. iommu=pt intel_iommu=on
2. intel_iommu=on
how about the scenario2

File: qemu/seabios_iommu.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    """
+    image_list = params.objects("images")
+    del image_list[0]
+    image_num = len(image_list)
+
+    error_context.context("Check data disks in monitor!", logging.info)
+    monitor_info_block = vm.monitor.info_block(False)
+    blocks = ','.join(monitor_info_block.keys())
+    for image in image_list:
+        if image not in blocks:
```
Comment (lolyu) [inline qemu/seabios_iommu.py]: why not directly check `if image not in monitor_info_block`?

File: qemu/seabios_iommu.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    """
+    Check guest data disks (except image1)
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    :param vm: VM object
+    :param session: VM session
+    """
+    image_list = params.objects("images")
```
Comment (lolyu) [inline qemu/seabios_iommu.py]: better to directly get the sublist via `image_list = params.objects("images")[1:]`

File: qemu/seabios_iommu.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            test.fail("drive_%s is missed: %s!" % (image, blocks))
+
+    error_context.context("Read and write on data disks!", logging.info)
+    os_type = params["os_type"]
+    if os_type == "linux":
+        sub_test_type = params.get("sub_test_type", "dd_test")
+        for image in image_list:
+            params["dd_if"] = "ZERO"
+            params["dd_of"] = image
+            utils_test.run_virt_sub_test(test, params, env, sub_test_type)
```
Comment (lolyu) [inline qemu/seabios_iommu.py]: I believe calling subtest is deprecated.

File: qemu/cfg/seabios_iommu.cfg (hunk 1 above)
Comment (ybduan) [reply qemu/cfg/seabios_iommu.cfg]: Modified, Thanks!

File: qemu/cfg/seabios_iommu.cfg (hunk 2 above)
Comment (ybduan) [reply qemu/cfg/seabios_iommu.cfg]: Confirmed with @lijinlijin , this is reasonable. Thanks!

File: qemu/cfg/seabios_iommu.cfg (hunk 3 above)
Comment (ybduan) [reply qemu/cfg/seabios_iommu.cfg]: Updated, Thanks!

File: qemu/seabios_iommu.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    """
+
+    if utils_misc.get_cpu_vendor(verbose=False) != 'GenuineIntel':
+        test.cancel("This case only support Intel platform")
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    login_timeout = int(params.get("login_timeout", 360))
+    session = vm.wait_for_login(timeout=login_timeout)
```
[the remark this answers is not in the document]
Comment (ybduan) [reply qemu/seabios_iommu.py]: This is not worked currently, change to close session as soon as it is useless. Thanks!

File: qemu/seabios_iommu.py (hunk 5 above)
Comment (ybduan) [reply qemu/seabios_iommu.py]: Updated, Thanks!

**Comment by ybduan** [comment]
(1/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.seabios_iommu.q35: PASS (129.94 s)
 (2/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.seabios_iommu.scenario1.q35: PASS (127.68 s)
 (3/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.seabios_iommu.scenario2.q35: PASS (81.49 s)
 (4/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.seabios_iommu.q35: PASS (467.82 s)

File: qemu/seabios_iommu.py (hunk 4 above)
Comment (ybduan) [reply qemu/seabios_iommu.py]: 'blocks' is useful in test fail info. Thanks!

**Comment by ybduan** [comment]
(1/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.seabios_iommu.q35: PASS (179.39 s)
 (2/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.seabios_iommu.scenario1.q35: PASS (263.34 s)
 (3/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.seabios_iommu.scenario2.q35: PASS (312.41 s)
 (4/4) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.seabios_iommu.q35: PASS (489.81 s)

File: qemu/seabios_iommu.py (hunk 6 above)
Comment (ybduan) [reply qemu/seabios_iommu.py]: Updated, using iozone instead. Could you please help review again? @lolyu Thanks!

**Comment by yanan-fu** [comment]
The three check point ``are not listed in polarion test case, why choose these check point?  
why combine three cases `virtio_blk`, `virtio_scsi`, `virtio_net`  together ?  

I guess you want to test `ping` to confirm `virtio_net`,  `Read and write data on data disks` and `info block` to confirm `virtio_scsi` and `virtio_blk`.   Split them is better.
 ```  
    2) Check 'info block'.
    3) Read and write data on data disks.
    4) Ping guest.
```

**Comment by ybduan** [comment]
It's no need to have 3 cases to test seabios with iommu, I have merge them to one case.

File: qemu/tests/cfg/seabios_iommu.cfg [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    extra_params = "-device intel-iommu,device-iotlb=on,intremap=on"
+    machine_type_extra_params = "kernel-irqchip=split"
+    virtio_dev_iommu_platform = on
+    virtio_dev_ats = on
+    images += " stg0 stg1"
+    image_name_stg0 = "images/stg0"
+    image_name_stg1 = "images/stg1"
+    image_size_stg0 = 1G
+    image_size_stg1 = 1G
+    data_image_size = 1G
```
Comment (luckyh) [inline qemu/tests/cfg/seabios_iommu.cfg]: What's the usage of this param? Why not directly use `image_size`?

File: qemu/tests/cfg/seabios_iommu.cfg [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    force_create_image_stg0 = yes
+    force_create_image_stg1 = yes
+    remove_image_stg0 = yes
+    remove_image_stg1 = yes
+    blk_extra_params_stg0 = "serial=TARGET_DISK0"
+    blk_extra_params_stg1 = "serial=TARGET_DISK1"
+    drive_format_stg0 = virtio
+    drive_format_stg1 = scsi-hd
+    virtio_scsi:
+        drive_bus_stg1 += 1
```
Comment (luckyh) [inline qemu/tests/cfg/seabios_iommu.cfg]: Would you please elaborate the usage of this line?

File: qemu/tests/seabios_iommu.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+    check_data_disks(test, params, env, vm, session)
+    session.close()
+
+    error_context.context("Ping guest!", logging.info)
+    guest_ip = vm.get_address()
+    status, output = utils_test.ping(guest_ip, count=10, timeout=20)
+    if status:
+        test.fail("Ping guest failed!")
+    elif utils_test.get_loss_ratio(output) == 100:
```
Comment (luckyh) [inline qemu/tests/seabios_iommu.py]: What if there is packet loss but the ratio doesn't reach 100%? Should it be a failure as well?

File: qemu/tests/seabios_iommu.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    2) Check 'info block'.
+    3) Read and write data on data disks.
+    4) Ping guest.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    if utils_misc.get_cpu_vendor(verbose=False) != 'GenuineIntel':
```
Comment (XueqiangWei) [inline qemu/tests/seabios_iommu.py]: get_cpu_vendor move to virttest.cpu. please use latest avocado codes. e.g. cpu.get_cpu_vendor()

File: qemu/tests/seabios_iommu.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+    error_context.context("Check data disks in monitor!", logging.info)
+    monitor_info_block = vm.monitor.info_block(False)
+    blocks = ','.join(monitor_info_block.keys())
+    for image in image_list:
+        if image not in blocks:
+            test.fail("drive_%s is missing: %s!" % (image, blocks))
+
+    error_context.context("Read and write data on data disks!", logging.info)
+    iozone_test = generate_instance(params, session, 'iozone')
```
Comment (XueqiangWei) [inline qemu/tests/seabios_iommu.py]: iozone = generate_instance(params, vm, 'iozone')

**Comment by XueqiangWei** [comment]
when test is failed, vm is not destroyed.
I think need add vm.destroy(gracefully=True) in the end

File: qemu/tests/cfg/seabios_iommu.cfg (hunk 8 above)
Comment (ybduan) [reply qemu/tests/cfg/seabios_iommu.cfg]: Updated, thanks!

File: qemu/tests/cfg/seabios_iommu.cfg (hunk 9 above)
Comment (ybduan) [reply qemu/tests/cfg/seabios_iommu.cfg]: Removed, thanks!

File: qemu/tests/seabios_iommu.py (hunk 10 above)
Comment (ybduan) [reply qemu/tests/seabios_iommu.py]: Updated, thanks!

File: qemu/tests/seabios_iommu.py (hunk 12 above)
Comment (ybduan) [reply qemu/tests/seabios_iommu.py]: Updated, thanks!

**Comment by ybduan** [comment]
(1/3) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.guest_iommu_test.q35: PASS (183.49 s)
 (2/3) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.guest_iommu_test.scenario1.q35: PASS (335.76 s)
 (3/3) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.guest_iommu_test.scenario2.q35: PASS (335.54 s)
 (1/1) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.guest_iommu_test.q35: PASS (327.07 s)

**Comment by ybduan** [comment]
@luckyh @XueqiangWei Please help review again, thanks!

File: qemu/tests/guest_iommu_test.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    2) Check 'info block'.
+    3) Read and write data on data disks.
+    4) Ping guest.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    if cpu.get_cpu_vendor(verbose=False) != 'GenuineIntel':
```
Comment (luckyh) [inline qemu/tests/guest_iommu_test.py]: Suggest using the one provided by `avocado.utils.cpu`

File: qemu/tests/guest_iommu_test.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+        return mount_points
+
+    image_list = params.objects("images")[1:]
+    image_num = len(image_list)
+
+    error_context.context("Check data disks in monitor!", logging.info)
+    monitor_info_block = vm.monitor.info_block(False)
+    blocks = ','.join(monitor_info_block.keys())
+    for image in image_list:
+        if image not in blocks:
```
Comment (luckyh) [inline qemu/tests/guest_iommu_test.py]: This approach is not always right, suppose that you have two images: `stg1` and `stg11`, the result would both be `False` while `blocks` is `"disk_stg11"`.

File: qemu/tests/guest_iommu_test.py (hunk 13 above)
Comment (ybduan) [reply qemu/tests/guest_iommu_test.py]: Updated, thanks!

File: qemu/tests/guest_iommu_test.py (hunk 14 above)
Comment (ybduan) [reply qemu/tests/guest_iommu_test.py]: Updated, thanks!

**Comment by ybduan** [comment]
(1/3) Host_RHEL.m8.ux.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.guest_iommu_test.q35: PASS (175.76 s)
 (2/3) Host_RHEL.m8.ux.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.guest_iommu_test.scenario1.q35: PASS (337.01 s)
 (3/3) Host_RHEL.m8.ux.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.guest_iommu_test.scenario2.q35: PASS (342.60 s)

**Comment by ybduan** [comment]
@luckyh Please help review again, thanks!

**Comment by XueqiangWei** [comment]
Get feedback from developer:  for issue "Guest os hang when setting interface with iommu config", no plans of fixing it on rhel7. And it's not necessary to test it on rhel7. So let's just test it on rhel8.

**Comment by ybduan** [comment]
Updated according to your comment, please help review again. Thanks! @XueqiangWei

**Comment by XueqiangWei** [comment]
(1/4) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.guest_iommu_test.q35: PASS (163.86 s)
 (2/4) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.guest_iommu_test.scenario1.q35: PASS (206.72 s)
 (3/4) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.guest_iommu_test.scenario2.q35: PASS (212.83 s)
 (4/4) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.guest_iommu_test.q35: PASS (257.64 s)

LGTM,
Acked-by xuwei@redhat.com

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**